### ISSUE TITLE
AMP shadow does not initialize components without user scrolling after visibilityState change from prerender to visible

### ISSUE THREAD

**Issue by p2er** [ampproject/amphtml#11983]
## What's the issue?

If multiple amp shadow documents are used underneath each other and should be set form visibilityState "prerender" to "visible" as the document is scrolled into view, the visibilityState change won't trigger AMP image and other components to load an render. 

## How do we reproduce the issue?

1. Open https://p2er.github.io/amp-articleslider/ampshadow-visibillityState
2. Make sure scroll location is at page top (scroll up and reload if not)
2. Click on "Slide3: prerender" link to jump to third amp shadow document
3. Click "Set slide 3 visible" button to change visibilityState to "visible"
4. Wait...  images, slider, etc won't initialize
5. Scroll up and down (wiggle) on the empty document, often scrolling back does the trick
6. Components initialize and render content 

## What browsers are affected?

Tested on Chrome and Safari 11.0.1, but most likely on all browsers. 

## Which AMP version is affected?

Powered by AMP ⚡ HTML shadows – Version 1509747505247

**Comment by aghassemi** [comment]
/to @cvializ 
/cc @dvdyakonov 

setting to P1.

**Comment by cvializ** [comment]
Able to reproduce as described. Will investigate

**Comment by cvializ** [comment]
It looks like this line is misbehaving: https://github.com/ampproject/amphtml/blob/feature/date-picker/src/service/resources-impl.js#L1430
```js
const shouldBeInViewport = (this.visible_ && r.isDisplayed() &&
    r.overlaps(visibleRect));
r.setInViewport(shouldBeInViewport);
```

Sometimes the `visibleRect` value appears to be negative, way above the top of the document, so `r.overlaps(visibleRect)` evaluated to `false`, so `r.setInViewport` is called with `false` when the ampdoc transitions from `prerender` to `visible`

**Comment by cvializ** [comment]
@jridgewell / @choumx any insight into why this might happen in a shadowdoc? Do you think it could have something to do with clicking a `href="#internal"` link to get to the document? I don't understand how the `visibleRect` could be above the top of the document. Unless I'm misunderstanding the values.

<img width="653" alt="screen shot 2017-11-14 at 3 18 46 pm" src="https://user-images.githubusercontent.com/2363700/32810284-2dc29dca-c94f-11e7-8a5e-1724e69819a7.png">

**Comment by ampprojectbot** [comment]
This is a high priority issue but it hasn't been updated in awhile. @cvializ Do you have any updates?

**Comment by ampprojectbot** [comment]
This is a high priority issue but it hasn't been updated in awhile. @cvializ Do you have any updates?

**Comment by ampprojectbot** [comment]
This is a high priority issue but it hasn't been updated in awhile. @cvializ Do you have any updates?

**Comment by ampprojectbot** [comment]
This is a high priority issue but it hasn't been updated in awhile. @cvializ Do you have any updates?

**Comment by ampprojectbot** [comment]
This is a high priority issue but it hasn't been updated in awhile. @cvializ Do you have any updates?

**Comment by ampprojectbot** [comment]
This issue seems to be in  Pending Triage for awhile. @cvializ Please triage this to an appropriate milestone.

**Comment by sparhami** [comment]
So it looks like `Viewport` has functionality to cache the value of scroll top (so it doesn't need to check it each time) here:

https://github.com/ampproject/amphtml/blob/2531b83833e02c6727633a9758ce4bb98b2e82bc/src/service/viewport/viewport-impl.js#L269-L274

What happens is that this function is called when trying to discover work when the AMP doc is attached. The value of scroll top is then calculated and cached. However, since the doc is not visible, the value of scroll top is not updated as the page scrolls. Once the page becomes visible, a scroll listener is attached here:

https://github.com/ampproject/amphtml/blob/2531b83833e02c6727633a9758ce4bb98b2e82bc/src/service/viewport/viewport-binding-natural.js#L74-L78

which then causes further reads of the value to be correct. I tried clearing the value of `scrollTop_` (and `scrollLeft_` when becoming visible, but the new calculated value appears to be incorrect.

**Comment by sparhami** [comment]
Spent some time looking into the code, but I don't think I have enough context to understand what is going on.

Here is what I have tried so far: https://github.com/ampproject/amphtml/commit/2e9018ebf6a514e0a14cbda33428708ff2ce3624

It still has a problem with `r.overlaps(visibleRect)` though.

It looks like a resource's `layoutBox_` takes into account the initial scroll position when the page is loaded, but in this example, does not appear to be updated as a result of toggling the document visibility. When it goes to check `r.overlaps(visibleRect)`, the visible rect has been updated taking into account the new scroll position, but the `layoutBox_` for each resource has not. I'm not sure if the `layoutBox_` is supposed to get updated to a new value or if the viewport rect is incorrect.

@jridgewell Any ideas on what might be going on here?
